k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+ *  the same macro for kernel could be used. See
+ *  `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * @code{.cpp}
+ * K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API());
+ * @endcode
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_API_SAFE_CALL(c) \
+      do {                           \
+        cudaError_t e = (c);         \
+        K2_CUDA_CHECK_ERROR(e);      \
+      } while (0)
+  #else
+    #define K2_CUDA_API_SAFE_CALL(c) (c)
+  #endif
+#endif
+
+/**
+ * @def K2_CUDA_KERNEL_SAFE_CALL([kernel|cuda_runtime_api])
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  Kernel launches do not return any error code, thus checking should after it.
+ *  To wait kernel to finish, cudaDeviceSynchronize is called between.
+ *  And to avoid pre-launch error disturbing, check error before kernel launch
+ *  is also needed.
+ *  (But not necessary if programmer promise each error is take cared properly.)
+ *
+ * @remark
+ *  macro `__VA_ARGS__` is used to pass the kernel<<<...>>> as one argument,
+ *  otherwise the compiler raises a error "passed 2 arguments, but takes just 1".
+ *
+ * @code{.cpp}
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_func<<<...>>>());
+ * @endcode
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_KERNEL_SAFE_CALL(...)               \
+      do {                                              \
+        (__VA_ARGS__);                                  \
+        K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
+        K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+      } while (0)
+  #else
+    #define K2_CUDA_KERNEL_SAFE_CALL(...) (__VA_ARGS__)
+  #endif
+#endif
+
+/**
+ * @def K2_DLOG
+ *
+ * @brief Log macro for printf statements.
+ *
+ * @details
+ *  `printf` is supported by both host and device. This Log is for debugging,
+ *  the error msg is printed to the stderr. The log msg always get printed,
+ *  regardless of macro `K2_MAKE_ERROR_CHECK`. Thus it should only be
+ *  make used of in debugging.
+ *
+ * @note
+ *  This code is refered from:
+ *  https://github.com/NVlabs/cub/blob/ \
+ *  c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/util_debug.cuh#L109
+ *
+ * @code{.cpp}
+ * K2_DLOG("Value is %d, string is %s ..", i, str);
+ * @endcode
+ */
+#if !defined(K2_DLOG)
+  #if !(defined(__clang__) && defined(__CUDA__))
+    #if (K2_PTX_ARCH == 0)
+      #define K2_DLOG(format, ...) printf(format,__VA_ARGS__)
+    #elif (K2_PTX_ARCH >= 200)
+      #define K2_DLOG(format, ...)                                            \
+        printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, blockIdx.x,  \
+               blockIdx.y, blockIdx.z, threadIdx.x, threadIdx.y, threadIdx.z, \
+               __VA_ARGS__)
+    #endif
+  #else
+    /**
+     * A hack to implement the variadic printf for clang,
```

Comment:
Why not to use `vprintf` or `vfprintf`? They are from the standard C library
and does not need special compiler support.